File: src2/MicrotonalHelper/equalTemperament.js
export function getEqualTemperamentNote(index, numberOfDivisions, period = 2) {
  if (numberOfDivisions <= 1) {
    throw Error("The numberOfDivisions must be positive and greater than 1");
  }
  if (period <= 1) {
    throw Error("The period must be positive and greater than 1");
  }
  return Math.pow(period, index / numberOfDivisions);
}

export function getEqualTemperamentScale(numberOfDivisions, period = 2) {
  const result = [];
  for (let index = 1; index <= numberOfDivisions; index++) {
    const note = getEqualTemperamentNote(index, numberOfDivisions, period);
    result.push(note);
  }
  return result;
}

export function getEqualTemperamentSubsets(
  numberOfNotes,
  numberOfDivisions,
  minDistanceBetweenNotes = 1,
  period = 2
) {
  if (numberOfNotes > numberOfDivisions) {
    throw Error(
      "The number of notes must be less than the number of divisions"
    );
  }

  if (minDistanceBetweenNotes < 1) {
    throw Error("The minDistanceBetweenNotes must be greater than 1");
  }

  if (numberOfNotes === numberOfDivisions) {
    //For this scenario, the full scale is the only result
    return {
      initialNumberOfCombinations: numberOfDivisions,
      numberOfCombinations: numberOfDivisions,
      result: [...new Array(numberOfDivisions).keys()]
    };
  }

  //N0 -> 0 to (numberOfDivisions - 1 - (numberOfNotes - 1))
  //N1 -> N0 + 1 to (numberOfDivisions - 1 - (numberOfNotes - 2))
  //N2 -> N1 + 1 to (numberOfDivisions - 1 - (numberOfNotes - 3))
  //NX -> N(X-1) to (numberOfDivisions - 1 - (numberOfNotes - (X+1)))
  // const result = [1]
  // let lastNoteIndex = 0;
  // for (let index = 1; index < numberOfNotes; index++) {

  //     const rangeMin = lastNoteIndex + 1
  //     const rangeMax =  (numberOfDivisions - 1 - (numberOfNotes - (index+1)))
  //     const randomNoteIndex = rangeMin + Math.floor(Math.random() * (rangeMax - rangeMin));
  //     const noteValue = getEqualTemperamentNote(randomNoteIndex, numberOfDivisions, period);
  //     result.push(noteValue);
  //     lastNoteIndex = randomNoteIndex;
  // }
  const bytesPerElement = Uint8Array.BYTES_PER_ELEMENT;
  const MAX_VALUE = Math.pow(2, bytesPerElement * 8) - 1;
  if (numberOfDivisions >= MAX_VALUE) {
    throw Error(`The number of divisions must be less than ${MAX_VALUE}`);
  }
  const initialNumberOfCombinations = getNumberOfCombinations(
    numberOfDivisions - 1,
    numberOfNotes - 1
  );
  const arrayLength = initialNumberOfCombinations * numberOfNotes;
  const MAX_ARRAY_LENGTH = Math.pow(2, 32) - 1;
  if (arrayLength > MAX_ARRAY_LENGTH) {
    throw Error(
      `Too much combinations to calculate: ${initialNumberOfCombinations.toLocaleString()}`
    );
  }
  const result = new Uint8Array(arrayLength);
  //const result = [];
  const data = recursive(
    0,
    1,
    numberOfNotes,
    numberOfDivisions,
    [0],
    result,
    0,
    minDistanceBetweenNotes
  );

  return {
    initialNumberOfCombinations,
    numberOfCombinations: data[1],
    result
  };
}

export function recursive(
  lastNoteIndex,
  level,
  numberOfNotes,
  numberOfDivisions,
  refArray,
  outputArray,
  outputIndex,
  minDistanceBetweenNotes
) {
  let size = 0;
  const rangeMin = lastNoteIndex + 1;
  const rangeMax = numberOfDivisions - 1 - (numberOfNotes - (level + 1));
  for (let index = rangeMin; index <= rangeMax; index++) {
    const lastRefIndex = refArray[refArray.length - 1];
    const indexDistance = index - lastRefIndex;
    if (indexDistance < minDistanceBetweenNotes) continue;

    const nextLevel = level + 1;
    const isLeaf = nextLevel === numberOfNotes;
    if (isLeaf) {
      for (let refIndex = 0; refIndex < refArray.length; refIndex++) {
        outputArray[outputIndex] = refArray[refIndex];
        outputIndex++;
      }
      outputArray[outputIndex] = index;
      outputIndex++;
      size++;
    } else {
      refArray.push(index);
      const tmpResult = recursive(
        index,
        nextLevel,
        numberOfNotes,
        numberOfDivisions,
        refArray,
        outputArray,
        outputIndex,
        minDistanceBetweenNotes
      );
      refArray.pop();
      outputIndex = tmpResult[0];
      size += tmpResult[1];
    }
  }

  return [outputIndex, size];
}

function factorial(num) {
  if (num < 0) return "Undefined";
  var fact = BigInt(1);
  for (var i = num; i > 1; i--) fact *= BigInt(i);
  return fact;
}
function getNumberOfCombinations(numberOfElements, groupSize) {
  return parseInt(
    factorial(numberOfElements) /
      (factorial(groupSize) * factorial(numberOfElements - groupSize))
  );
}

File: src3/MicrotonalExplorer/MicrotonalHelpers/EqualTemperament.cs
using System.Numerics;

namespace MicrotonalExplorer;

public record EqualTemperamentSubsetResult
(
    uint InitialNumberOfCombinations,
    uint NumberOfCombinations,
    int[] Result
);


public static class EqualTemperament
{
    public static EqualTemperamentSubsetResult GetEqualTemperamentSubsets(
      int numberOfNotes,
      int numberOfDivisions,
      int minDistanceBetweenNotes = 1
    )
    {
        if (numberOfNotes > numberOfDivisions)
        {
            throw new Exception(
              "The number of notes must be less than the number of divisions"
            );
        }

        if (minDistanceBetweenNotes < 1)
        {
            throw new Exception("The minDistanceBetweenNotes must be greater than 1");
        }

        if (numberOfNotes == numberOfDivisions)
        {
            //For this scenario, the full scale is the only result
            return new EqualTemperamentSubsetResult(
                (uint)numberOfDivisions,
                (uint)numberOfDivisions,
                Enumerable.Range(0, numberOfDivisions).ToArray()
            );
        }

        if (numberOfDivisions >= int.MaxValue)
        {
            throw new Exception($"The number of divisions must be less than ${int.MaxValue}");
        }
        var initialNumberOfCombinations = GetNumberOfCombinations(
          numberOfDivisions - 1,
          numberOfNotes - 1
        );
        var arrayLength = initialNumberOfCombinations * numberOfNotes;
        if (arrayLength >= uint.MaxValue)
        {
            throw new Exception($"Too much combinations to calculate: ${initialNumberOfCombinations.ToString("N")}");
        }
        var result = new int[arrayLength];
        var refArray = new int[numberOfNotes];
        var refArrayIndex = 0;
        refArray[refArrayIndex] = 0;

        var data = Recursive(
          0,
          1,
          numberOfNotes,
          numberOfDivisions,
          refArrayIndex,
          refArray,
          result,
          0,
          minDistanceBetweenNotes
        );

        return new EqualTemperamentSubsetResult(
            initialNumberOfCombinations,
            data.size,
            result
        );
    }

    static (int outputIndex, uint size) Recursive(
      int lastNoteIndex,
      int level,
      int numberOfNotes,
      int numberOfDivisions,
      int refArrayIndex,
      int[] refArray,
      int[] outputArray,
      int outputIndex,
      int minDistanceBetweenNotes
    )
    {
        uint size = 0;
        var rangeMin = lastNoteIndex + 1;
        var rangeMax = numberOfDivisions - 1 - (numberOfNotes - (level + 1));
        for (var index = rangeMin; index <= rangeMax; index++)
        {
            var lastRefIndex = refArray[refArrayIndex];
            var indexDistance = index - lastRefIndex;
            if (indexDistance < minDistanceBetweenNotes) continue;

            var nextLevel = level + 1;
            var isLeaf = nextLevel == numberOfNotes;
            if (isLeaf)
            {
                for (var refIndex = 0; refIndex <= refArrayIndex; refIndex++)
                {
                    outputArray[outputIndex] = refArray[refIndex];
                    outputIndex++;
                }
                outputArray[outputIndex] = index;
                outputIndex++;
                size++;
            }
            else
            {
                refArrayIndex++;
                refArray[refArrayIndex] = index;
                var tmpResult = Recursive(
                  index,
                  nextLevel,
                  numberOfNotes,
                  numberOfDivisions,
                  refArrayIndex,
                  refArray,
                  outputArray,
                  outputIndex,
                  minDistanceBetweenNotes
                );
                refArray[refArrayIndex] = 0;
                refArrayIndex--;
                outputIndex = tmpResult.outputIndex;
                size += tmpResult.size;
            }
        }

        return (outputIndex, size);
    }

    static BigInteger Factorial(int num)
    {
        if (num < 0) throw new Exception("Undefined");
        BigInteger fact = 1;
        for (var i = num; i > 1; i--){
            fact *= i;
        }
        return fact;
    }

    static uint GetNumberOfCombinations(int numberOfElements, int groupSize)
    {
        var result = Factorial(numberOfElements) /
            (Factorial(groupSize) * Factorial(numberOfElements - groupSize));
        return (uint)result;
    }

    public static float GetEqualTemperamentNote(int noteIndex, int numberOfDivisions, float period)
    {
        if (numberOfDivisions <= 1)
        {
            throw new Exception("The numberOfDivisions must be positive and greater than 1");
        }
        if (period <= 1)
        {
            throw new Exception("The period must be positive and greater than 1");
        }
        return (float)Math.Pow(period, (float)noteIndex / numberOfDivisions);
    }
}
